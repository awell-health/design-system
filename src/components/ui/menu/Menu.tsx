import { cloneElement } from 'react';
import { cn } from '../../../lib/utils';
import { MenuItem } from './types';

export interface Props extends React.HTMLAttributes<HTMLUListElement> {
  items: MenuItem[];
}

function Menu({ className, items, ...props }: Props) {
  const renderMenItem = (item: MenuItem) => {
    const { icon, badge, children, className = '' } = item;

    const itemClassNames = cn(
      'group flex box-border',
      'h-9 pl-2 pr-3 py-2 rounded-md text-sm font-medium text-slate-700',
      'hover:border-blue-100 hover:bg-blue-100 hover:text-blue-700',
      'active:!border-blue-100 active:!bg-blue-50 active:!text-blue-700',
      !icon && 'ml-3',
      className
    );

    const iconComponent = cloneElement(icon ?? <></>, {
      className: 'fill-slate-500 group-hover:fill-blue-700 group-active:fill-blue-700'
    });

    const label = (
      <span className='flex flex-1 items-center gap-3'>
        {icon && <span>{iconComponent}</span>}
        {item.label}
        {badge && <span className={cn(children && 'mr-1')}>{badge}</span>}
      </span>
    );
    return (
      <li>
        {!children && <a className={itemClassNames}>{label}</a>}
        {children && (
          <details>
            <summary className={itemClassNames}>{label}</summary>
            <ul className='before:w-0'>{children.map(renderMenItem)}</ul>
          </details>
        )}
      </li>
    );
  };

  return (
    <ul className={cn('menu p-0', className)} {...props}>
      {items.map(renderMenItem)}
    </ul>
  );
}

export { Menu };
