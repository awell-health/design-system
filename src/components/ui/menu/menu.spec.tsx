import { fireEvent, render } from '@testing-library/react';
import { expect, it, describe } from 'vitest';
import { Menu } from './Menu';

describe('Menu', () => {
  const subject = (props = {}) => {
    const defaultProps = {
      items: [
        {
          label: 'Single Item'
        },
        {
          label: 'Parent Item',
          children: [
            {
              label: 'First child'
            },
            {
              label: 'Second child',
              children: [{ label: 'Second level child' }]
            }
          ]
        }
      ]
    };
    return render(<Menu {...defaultProps} {...props} />);
  };

  it('match snapshot', () => {
    expect(subject().container).toMatchSnapshot();
  });

  it('renders icon', () => {
    const { getByText } = subject({
      items: [{ label: 'Test', icon: <div>icon</div> }]
    });

    expect(getByText('icon')).toBeVisible();
  });

  it('renders badge', () => {
    const { getByText } = subject({
      items: [{ label: 'Test', badge: <span>badge</span> }]
    });

    expect(getByText('badge')).toBeVisible();
  });

  it('opens children menu', () => {
    const { getByText } = subject();

    expect(getByText('First child')).not.toBeVisible();

    fireEvent.click(getByText('Parent Item'));

    expect(getByText('First child')).toBeVisible();
  });

  it('opens second level children menu', () => {
    const { getByText } = subject();

    expect(getByText('Second level child')).not.toBeVisible();

    fireEvent.click(getByText('Parent Item'));
    fireEvent.click(getByText('Second child'));

    expect(getByText('Second level child')).toBeVisible();
  });
});
