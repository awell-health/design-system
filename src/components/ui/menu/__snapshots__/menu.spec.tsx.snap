// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Menu > match snapshot 1`] = `
<div>
  <ul
    class="menu p-0"
  >
    <li>
      <a
        class="group flex box-border h-9 pl-2 pr-3 py-2 rounded-md text-sm font-medium text-slate-700 hover:border-blue-100 hover:bg-blue-100 hover:text-blue-700 active:!border-blue-100 active:!bg-blue-50 active:!text-blue-700 ml-3"
      >
        <span
          class="flex flex-1 items-center gap-3"
        >
          Single Item
        </span>
      </a>
    </li>
    <li>
      <details>
        <summary
          class="group flex box-border h-9 pl-2 pr-3 py-2 rounded-md text-sm font-medium text-slate-700 hover:border-blue-100 hover:bg-blue-100 hover:text-blue-700 active:!border-blue-100 active:!bg-blue-50 active:!text-blue-700 ml-3"
        >
          <span
            class="flex flex-1 items-center gap-3"
          >
            Parent Item
          </span>
        </summary>
        <ul
          class="before:w-0"
        >
          <li>
            <a
              class="group flex box-border h-9 pl-2 pr-3 py-2 rounded-md text-sm font-medium text-slate-700 hover:border-blue-100 hover:bg-blue-100 hover:text-blue-700 active:!border-blue-100 active:!bg-blue-50 active:!text-blue-700 ml-3"
            >
              <span
                class="flex flex-1 items-center gap-3"
              >
                First child
              </span>
            </a>
          </li>
          <li>
            <details>
              <summary
                class="group flex box-border h-9 pl-2 pr-3 py-2 rounded-md text-sm font-medium text-slate-700 hover:border-blue-100 hover:bg-blue-100 hover:text-blue-700 active:!border-blue-100 active:!bg-blue-50 active:!text-blue-700 ml-3"
              >
                <span
                  class="flex flex-1 items-center gap-3"
                >
                  Second child
                </span>
              </summary>
              <ul
                class="before:w-0"
              >
                <li>
                  <a
                    class="group flex box-border h-9 pl-2 pr-3 py-2 rounded-md text-sm font-medium text-slate-700 hover:border-blue-100 hover:bg-blue-100 hover:text-blue-700 active:!border-blue-100 active:!bg-blue-50 active:!text-blue-700 ml-3"
                  >
                    <span
                      class="flex flex-1 items-center gap-3"
                    >
                      Second level child
                    </span>
                  </a>
                </li>
              </ul>
            </details>
          </li>
        </ul>
      </details>
    </li>
  </ul>
</div>
`;
